---

_You are currently looking at **version 1.2** of this notebook. To download notebooks and datafiles, as well as get help on Jupyter notebooks in the Coursera platform, visit the [Jupyter Notebook FAQ](https://www.coursera.org/learn/python-social-network-analysis/resources/yPcBs) course resource._

---

# Assignment 4

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import networkx as nx
import pandas as pd
import numpy as np
import pickle

---

## Part 1 - Random Graph Identification

For the first part of this assignment you will analyze randomly generated graphs and determine which algorithm created them.

In [5]:
P1_Graphs = pickle.load(open('/content/drive/MyDrive/A4_graphs','rb'))
P1_Graphs

<br>
`P1_Graphs` is a list containing 5 networkx graphs. Each of these graphs were generated by one of three possible algorithms:
* Preferential Attachment (`'PA'`)
* Small World with low probability of rewiring (`'SW_L'`)
* Small World with high probability of rewiring (`'SW_H'`)

Anaylze each of the 5 graphs and determine which of the three algorithms generated the graph.

*The `graph_identification` function should return a list of length 5 where each element in the list is either `'PA'`, `'SW_L'`, or `'SW_H'`.*

In [10]:
def graph_identification():
    algos = dict()
    for G in P1_Graphs:
        deg = G.degree()
        degree_hist = [list(
                           deg.values()
                           ).count(i) / float(nx.number_of_nodes(G)) 
                           for i in sorted(set(deg.values()))]
        avg_clustering = nx.average_clustering(G)

        if len(degree_hist) > 10:
            algos[G] = "PA"
        elif avg_clustering < 0.1:
            algos[G] = "SW_H"
        else:
            algos[G] = "SW_L"

    return list(algos.values())

---

## Part 2 - Company Emails

For the second part of this assignment you will be workking with a company's email network where each node corresponds to a person at the company, and each edge indicates that at least one email has been sent between two people.

The network also contains the node attributes `Department` and `ManagementSalary`.

`Department` indicates the department in the company which the person belongs to, and `ManagementSalary` indicates whether that person is receiving a management position salary.

In [12]:
G = nx.read_gpickle('/content/drive/MyDrive/email_prediction.txt')

# print(nx.info(G))

### Part 2A - Salary Prediction

Using network `G`, identify the people in the network with missing values for the node attribute `ManagementSalary` and predict whether or not these individuals are receiving a management position salary.

To accomplish this, you will need to create a matrix of node features using networkx, train a sklearn classifier on nodes that have `ManagementSalary` data, and predict a probability of the node receiving a management salary for nodes where `ManagementSalary` is missing.



Your predictions will need to be given as the probability that the corresponding employee is receiving a management position salary.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 252 with the data being the probability of receiving management salary, and the index being the node id.

    Example:
    
        1       1.0
        2       0.0
        5       0.8
        8       1.0
            ...
        996     0.7
        1000    0.5
        1001    0.0
        Length: 252, dtype: float64

In [13]:
def salary_predictions():
    from sklearn.preprocessing import MinMaxScaler
    from sklearn.neural_network import MLPClassifier
    

    df = pd.DataFrame(index=G.nodes())
    # adding features to df
    df["clustering"] = pd.Series(nx.clustering(G))
    df["degree"] = pd.Series(nx.degree(G))
    df["degree_centrality"] = pd.Series(nx.degree_centrality(G))
    df["closeness"] = pd.Series(nx.closeness_centrality(G, normalized=True))
    df["betweeness"] = pd.Series(nx.betweenness_centrality(G, normalized=True))
    df["pagerank"] = pd.Series(nx.pagerank(G))
    ms = nx.get_node_attributes(G, "ManagementSalary").values()
    df["management"] = pd.Series(list(ms))

    # train and test split
    train = df[~pd.isnull(df["management"])]
    test = df[pd.isnull(df["management"])]

    X_train = train.iloc[:, :-1]
    y_train = train.iloc[:, -1]
    X_test = test.iloc[:, :-1]

    # normalization
    scaler = MinMaxScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)

    # train the model
    mlp = MLPClassifier(hidden_layer_sizes=[10, 5], 
                        alpha=5, 
                        random_state=0, 
                        solver="lbfgs").fit(X_train_scaled, y_train)
    
    return pd.Series(mlp.predict_proba(X_test_scaled)[:, 1],
                     X_test.index)

### Part 2B - New Connections Prediction

For the last part of this assignment, you will predict future connections between employees of the network. The future connections information has been loaded into the variable `future_connections`. The index is a tuple indicating a pair of nodes that currently do not have a connection, and the `Future Connection` column indicates if an edge between those two nodes will exist in the future, where a value of 1.0 indicates a future connection.

In [15]:
future_connections = pd.read_csv('/content/drive/MyDrive/Future_Connections.csv', index_col=0, converters={0: eval})
future_connections.head(10)

Future Connection
(6, 840)                  0.0
(4, 197)                  0.0
(620, 979)                0.0
(519, 872)                0.0
(382, 423)                0.0
(97, 226)                 1.0
(349, 905)                0.0
(429, 860)                0.0
(309, 989)                0.0
(468, 880)                0.0

Using network `G` and `future_connections`, identify the edges in `future_connections` with missing values and predict whether or not these edges will have a future connection.

To accomplish this, you will need to create a matrix of features for the edges found in `future_connections` using networkx, train a sklearn classifier on those edges in `future_connections` that have `Future Connection` data, and predict a probability of the edge being a future connection for those edges in `future_connections` where `Future Connection` is missing.



Your predictions will need to be given as the probability of the corresponding edge being a future connection.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 122112 with the data being the probability of the edge being a future connection, and the index being the edge as represented by a tuple of nodes.

    Example:
    
        (107, 348)    0.35
        (542, 751)    0.40
        (20, 426)     0.55
        (50, 989)     0.35
                  ...
        (939, 940)    0.15
        (555, 905)    0.35
        (75, 101)     0.65
        Length: 122112, dtype: float64

In [ ]:
def new_connections_predictions():
    from sklearn.ensemble import GradientBoostingClassifier
    

    # adding features to df
    future_connections["pref_attch"] = [
        list(nx.preferential_attachment(G, [nodes]))[0][2] 
        for nodes in future_connections.index]

    future_connections["common_neighbors"] = [
        len(list(nx.common_neighbors(G, nodes[0], nodes[1])))
        for nodes in future_connections.index]
    
    # train test split
    train = future_connections[~future_connections["Future Connection"].isnull()]
    test = future_connections[future_connections["Future Connection"].isnull()]
    X_train = train[["pref_attch", "common_neighbors"]].values
    y_train = train["Future Connection"].values
    X_test = test[["pref_attch", "common_neighbors"]].values

    # train the model
    gb = GradientBoostingClassifier().fit(X_train, y_train)

    return pd.Series(gb.predict_proba(X_test)[:, 1], index=test.index)